File: csharp/src/Vista.SDK/Internal/LocationParsingErrorBuilder.cs
namespace Vista.SDK.Internal;

internal sealed record LocationParsingErrorBuilder
{
    private readonly List<(LocationValidationResult name, string message)> _errors;

    internal static readonly LocationParsingErrorBuilder Empty = new();

    internal LocationParsingErrorBuilder() =>
        _errors = new List<(LocationValidationResult, string)>();

    internal LocationParsingErrorBuilder AddError(
        LocationValidationResult validationResult,
        string message
    )
    {
        _errors.Add((validationResult, message));
        return this;
    }

    internal bool HasError => _errors.Count > 0;

    internal static LocationParsingErrorBuilder Create() => new();

    internal ParsingErrors Build() =>
        _errors.Count == 0
            ? ParsingErrors.Empty
            : new ParsingErrors(_errors.Select((t, m) => (t.name.ToString(), t.message)).ToArray());
}

File: cpp/src/SDK/core/internal/parsing/LocationParsingErrorBuilder.cpp
#include "LocationParsingErrorBuilder.h"

#include "dnv/vista/sdk/core/ParsingErrors.h"

namespace dnv::vista::sdk::internal
{
    bool LocationParsingErrorBuilder::hasError() const noexcept
    {
        return !m_errors.empty();
    }

    ParsingErrors LocationParsingErrorBuilder::build() const
    {
        if (m_errors.empty())
        {
            return ParsingErrors{};
        }

        std::vector<ParsingErrors::ErrorEntry> errors;
        errors.reserve(m_errors.size());

        for (const auto& [result, message] : m_errors)
        {
            std::string resultStr;
            switch (result)
            {
                case LocationValidationResult::Invalid:
                {
                    resultStr = "Invalid";
                    break;
                }
                case LocationValidationResult::InvalidCode:
                {
                    resultStr = "InvalidCode";
                    break;
                }
                case LocationValidationResult::InvalidOrder:
                {
                    resultStr = "InvalidOrder";
                    break;
                }
                case LocationValidationResult::NullOrWhiteSpace:
                {
                    resultStr = "NullOrWhiteSpace";
                    break;
                }
                case LocationValidationResult::Valid:
                {
                    resultStr = "Valid";
                    break;
                }
            }

            errors.emplace_back(std::move(resultStr), message);
        }

        return ParsingErrors{ std::move(errors) };
    }

    LocationParsingErrorBuilder& LocationParsingErrorBuilder::addError(
        LocationValidationResult validationResult, std::string_view message)
    {
        m_errors.emplace_back(validationResult, std::string{ message });
        return *this;
    }
} // namespace dnv::vista::sdk::internal
